### PR TITLE
Shorthand supports for dash-core-components

### PR DESCRIPTION
**Pull request by workaholicpanda**
_(this PR replaces PR #1747 with one squashed commit)_

Implements [this issue](https://github.com/plotly/dash-core-components/issues/960) to support optional shorthand supports for below components.
You can compare [old format](https://gist.github.com/jackparmer/864545d0c7f7596ed6b4c9b18a3b252c) and [new format](https://gist.github.com/workaholicpanda/e54d762fd8ae6e755eebfbef40c42bec).

## Dependency Update
 - [x] `Input` `Output` `State` can now accept component object as well as string value of the component key
    - [x] (Required) A separate PR will need to be merged before this

## Updated Components with change notes
 - Dropdown
   - [x] `options` props can now be passed without being a keyword argument
   - [x] `options` can now accept array of string as well as previous format (array of dictionary)

 - RangeSlider
   - [x] Changed props order to make it start with `min` `max` `step` 

 - Checklist
   - [x] `options` props becomes first argument, can be passed without keyword
   - [x] Added `inline` props to specify if `options` are array of string or dictionary (previous format)

- RadioItems
   - [x] `options` props becomes first argument, can be passed without keyword
   - [x] Added `inline` props to specify if `options` are array of string or dictionary (previous format)

- DataTable
  - [x] `data` and `columns` props becomes first 2 arguments, can be passed without keyword
  - [x] (Optional) Separate PR will be followed to extract `columns` automatically from given `data`

### CODE REVIEW

**Comment by workaholicpanda** [comment]
# Changelog

## Dash and Dash Renderer
 - Any `Component` used as `Dependency` reference will have auto generated `id` if not given
 - `Input` `Output` `State` can now accept component object as well as string value of the component key

## Dash Core Components
### Shorthand supports & Behavioral changes
  - `Dropdown`:
    - **Props order changed** - `options` `value` can be passed as positional arguments without keyword
    - **Props type changed** - Additional type supports for `options` - array of `string|number|bool` where `label` and `value` are inferred.
  - `RangeSlider`, `Slider`
    - **Props order changed** - `min` `max` `step` can be passed as positional arguments without keyword 
    - **Props type changed** - Additional type supports for `options` - array of `string|number|bool` where `label` and `value` are inferred.
    - **Props behavior changed**
      - `step` will be calculated implicitly if not given.
      - `marks` will be auto generated if not given, according to `min` and `max` respecting `step` as well if given. Auto generated marks labels are SI unit formatted using [d3-format](https://github.com/d3/d3-format#locale_formatPrefix)
      - Specifying `marks=None` will prevent auto generating marks.
  - `Checklist`, `RadioItems`
    - **Props order changed** - `options` props can now be passed as positional arguments without keyword
    - **Props type changed** - Additional type supports for `options` - array of `string|number|bool` where `label` and `value` are inferred.
    -  **Props added** - New props `inline:Boolean` to set label style `display: 'inline-block' when True is given
  - `DataTable`
    - **Props order changed** - `data` and `columns` props becomes the first 2 arguments, and now can be passed as positional arguments without keyword 
    - **Props behavior changed** - `columns` is now extracted automatically from the given `data`
    - **Props type changed** - `data` supports `bool` items in array

File: components/dash-core-components/tests/integration/sliders/test_sliders_shorthands.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            [
+                html.Div(
+                    [
+                        html.Div(
+                            f"min={min}, max={max}(=10^{n})",
+                            style={"marginBottom": 15, "marginTop": 25},
+                        ),
+                        dcc.Slider(min, max),
+                        dcc.RangeSlider(min, max),
+                    ]
```
Comment (chriddyp) [inline components/dash-core-components/tests/integration/sliders/test_sliders_shorthands.py:170]:
```suggestion
        max = pow(10, n)
        
        LAYOUT.extend(
            [
                html.Div(
                    [
                       
                        html.B(
                            f"min={min}, max={max}(=10^{n})",
                            style={"marginBottom": 15, "marginTop": 25},
                        ),
                        (html.Div(
                            '(Known issue: Slider does not seem to work for precision below 10^(-6))'
                        ) if n <= 6 else None),
                        html.Div('value is undefined'),
                        dcc.Slider(min, max),
                        dcc.RangeSlider(min, max),
                        
                        html.Div(f'value=0.4 * 10^{n}'),
                        dcc.Slider(min, max, value=0.4*max),
                        dcc.RangeSlider(min, max, value[0.2 * max, 0.4 * max]),

                        html.Div(f'value=0.5 * 10^{n}'),
                        dcc.Slider(min, max, value=0.4*max),
                        dcc.RangeSlider(min, max, value[0.2 * max, 0.5 * max]),
                    ]
```

File: tests/unit/development/test_base_component.py [hunk 2]
```diff
@@ -473,3 +474,37 @@ def test_debc027_component_error_message():
         + "keyword argument: `asdf`\n"
         + "Allowed arguments: {}".format(", ".join(sorted(html.Div()._prop_names)))
     )
+
+
+def test_set_random_id():
+    app = Dash(__name__)
+
+    input1 = dcc.Input(value="Hello")
+    input2 = dcc.Input(value="Hello")
```
Comment (chriddyp) [inline tests/unit/development/test_base_component.py:483]:
```suggestion
    input1 = dcc.Input(value="Hello Input 1")
    input2 = dcc.Input(value="Hello Input 2")
```

Just to keep things unambiguous

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,148 @@
...
+    * `µ` - micro, 10⁻⁶
+    * `m` - milli, 10⁻³
+    * `​` (none) - 10⁰
+    * `k` - kilo, 10³
+    * `M` - mega, 10⁶
+    * `G` - giga, 10⁹
+    * `T` - tera, 10¹²
+    * `P` - peta, 10¹⁵
+    * `E` - exa, 10¹⁸
+    * `Z` - zetta, 10²¹
```
Comment (chriddyp) [inline CHANGELOG.md:119]:
```suggestion
    The SI units and ranges supported in `marks` are:

    * `µ` - micro, 10⁻⁶
    * `m` - milli, 10⁻³
    * `​` (none) - 10⁰
    * `k` - kilo, 10³
    * `M` - mega, 10⁶
    * `G` - giga, 10⁹
    * `T` - tera, 10¹²
    * `P` - peta, 10¹⁵
    * `E` - exa, 10¹⁸
    
    _Ranges below 10µ are not supported by the Slider. This is a bug: https://github.com/plotly/dash/issues/1766_
```

**Review by chriddyp — APPROVED** [review]
💃🏼 Great work! Let's do it.

**Pull request merged**